Emit timestamps on each line of output

karkhaz opened the pull request:
This commit adds the --timestamp command line option to allow cprover
tools to emit timestamps before each line of output. The timestamps are
either pretty-formatted wall-clock times (which may decrease as time
passes in the event of a leap-second or daylight savings time event), or
raw numbers that are guaranteed to increase. In either case, the
timestamps have microsecond precision if supported by the underlying
platform.

smowton commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,54 @@
+class monotonic_timestampert : public timestampert
+{
+public:
+  virtual const std::string stamp(const std::string &message) const override;
```
I believe `const` on a by-value return type doesn't really mean anything

smowton commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,54 @@
+" --timestamp <monotonic|wall> print microsecond-precision timestamps.\n" \
+"                              monotonic: timestamps are relative to an\n" \
+"                              arbitrary start point, guaranteed to " \
+"increase,\n" \
```
fix this formatting

smowton commented on src/util/ui_message.cpp:
```diff
@@ -76,6 +87,7 @@ ui_message_handlert::~ui_message_handlert()
   case uit::PLAIN:
     break;
   }
+  delete time;
```
Use a `unique_ptr` rather than hand-pairing `new` and `delete`

smowton reviewed (commented):
Seems reasonable; some suggestions, and I suggest rather than have a `stamp` method that takes a message and returns it with something prepended, just provide a `current_time_string` method that returns the current time and have `messaget` do the concatenation.

martin-cs commented on src/cbmc/cbmc_parse_options.h:
```diff
@@ -70,7 +72,7 @@ class optionst;
   "(fixedbv)(floatbv)(all-claims)(all-properties)" // legacy, and will eventually disappear // NOLINT(whitespace/line_length)
 
-class cbmc_parse_optionst:
+class cbmc_parse_optionst :
```
Please avoid pure whitespace changes.

martin-cs commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -82,7 +84,7 @@ Author: Daniel Kroening, kroening@kroening.com
 
 
-class goto_instrument_parse_optionst:
+class goto_instrument_parse_optionst :
```
Ditto.

martin-cs reviewed (commented):
I can see how this would be useful but I'm a little concerned about the duplication of functionality between timestamper.h and our existing time.h.  Doesn't this give two different timing APIs?

karkhaz commented:
@smowton thanks, I've applied your suggestions.
@martin-cs thanks, I've applied your suggestions. I propose that we remove time.h and time_stopping.h entirely, since these were written in 2004 before the C++11 time API existed. I've updated the patchset to remove the old time API and make all clients of the old API use C++11 std::chrono. Thoughts?

smowton commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,82 @@
+/// This class hierarchy supports generation of timestamps in various
+/// textual formats. The timestamps returned by instances of this class
+/// are empty; more meaningful timestamps are returned by derived classes.
+class timestampert
```
Clarify whether this abstract, or is instantiating the base useful / sensible?

karkhaz replied:
This is not abstract, it is instantiated when `--timestamps` is not passed on the command line. This means that ui_messaget can always call `time->stamp()` without bothering to check whether `--timestamps` was passed; virtual dispatch takes care of whether to emit an actual timestamp or just the empty string. Should I clarify this in comments?

smowton replied:
Yes please, it looks at the moment like an abstract base that provides a default implementation for a  concrete subclass

smowton reviewed (approved):
LGTM

tautschnig commented on src/cbmc/all_properties.cpp:
```diff
@@ -56,8 +57,9 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
-  absolute_timet sat_start=current_time();
+  std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>
+    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(
+      std::chrono::steady_clock::now());
```
Would it be worthwhile keeping a `time_stopping.h` (but really just the header) with

````c++
#include <chrono>

using absolute_timet = 
  std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>;

absolute_timet current_time()
{
  return
    std::chrono::time_point_cast<std::chrono::seconds(
      std::chrono::steady_clock::now());
}
````

tautschnig commented on .gitignore:
```diff
@@ -45,7 +45,7 @@ src/ansi-c/gcc_builtin_headers_tm.inc
 src/ansi-c/gcc_builtin_headers_power.inc
 src/ansi-c/gcc_builtin_headers_ubsan.inc
-src/java_bytecode/java_core_models.inc
+src/java_bytecode/library/java_core_models.inc
```
For future reference: please put those changes in a separate PR as it is trivial to merge (and would avoid others accidentally coming up with the same).

karkhaz replied:
Acknowledged, will do this next time!

tautschnig commented on src/cbmc/all_properties.cpp:
```diff
@@ -57,9 +57,10 @@ safety_checkert::resultt bmc_all_propertiest::operator()()
   std::chrono::time_point<std::chrono::steady_clock, std::chrono::seconds>
-    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(
-      std::chrono::steady_clock::now());
+    sat_start = std::chrono::time_point_cast<std::chrono::seconds>(now);
```
Is this change (and the one above) possibly in the wrong commit?

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1029,6 +1029,7 @@ void cbmc_parse_optionst::help()
     " --json-ui                    use JSON-formatted output\n"
     HELP_GOTO_TRACE
     " --verbosity #                verbosity level\n"
+    HELP_TIMESTAMP
```
Please add to goto-analyzer, goto-diff as well.

martin-cs replied:
+1 that would be lovely.

karkhaz replied:
This is done

tautschnig commented on src/util/timestamper.cpp:
```diff
@@ -0,0 +1,84 @@
+#include <cstdlib>
+
+#include "invariant.h"
+#include "timestamper.h"
```
Nit pick: I think we've move to including the header file associated with an implementation to the top of the list of includes.

tautschnig commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,95 @@
+
+
+#include <memory>
+#include <sstream>
```
Where is this used/required?

tautschnig commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,95 @@
+    /// wall_clock_timestampert
+    WALL_CLOCK
+  };
+  timestampert() {}
```
Seems unnecessary as it is implicit?

tautschnig commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,95 @@
+    WALL_CLOCK
+  };
+  timestampert() {}
+  virtual ~timestampert() {}
```
I'd suggest `virtual ~timestampert() = default;`

tautschnig commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,95 @@
+  virtual ~timestampert() {}
+
+  /// \brief Default timestamp: the empty string
+  virtual std::string stamp() const
```
It seems you'd need `#include <string>` rather than `<sstream>`?

tautschnig commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,95 @@
+};
+
+#endif
+#endif
```
Nit picks: add `/* CPROVER_UTIL_TIMESTAMPER_H */` in the very last line, and swap the blank line and the `#endif` before.

tautschnig commented on src/util/timestamper.cpp:
```diff
@@ -0,0 +1,84 @@
+  auto cnt = time_stamp.time_since_epoch().count();
+  std::div_t divmod = div(cnt, 1000000);
+  char u_seconds_fmt[7];
+  std::snprintf(u_seconds_fmt, sizeof(u_seconds_fmt), "%06d", divmod.rem);
```
How about `std::setfill('0') << std::setw(6) << divmod.rem` to get rid of `snprintf`?

tautschnig commented on src/util/timestamper.cpp:
```diff
@@ -0,0 +1,84 @@
+  std::tm *local = std::localtime(&tt);
+  char time_buf[WALL_TEXT_BUFFER_LENGTH];
+  int chars =
+    std::strftime(time_buf, WALL_TEXT_BUFFER_LENGTH, WALL_FORMAT, local);
```
You might use `std::put_time` instead?

tautschnig commented on src/util/ui_message.cpp:
```diff
@@ -57,7 +61,19 @@ ui_message_handlert::ui_message_handlert(
+      : timestampert::clockt::NONE
+#else
+      timestampert::clockt::NONE
+#endif
```
I don't think you need the `#ifndef _WIN32` here: for `_WIN32`, your code simply ignores the `clock_type` argument, and hence you're completely safe to us the full ternary branch above in either case!?

tautschnig reviewed (commented):
No blockers, but a few ideas/suggestions.

karkhaz committed:
Emit timestamps on each line of output

This commit adds the --timestamp command line option to allow cprover
tools to emit timestamps before each line of output. The timestamps are
either pretty-formatted wall-clock times (which may decrease as time
passes in the event of a leap-second or daylight savings time event), or
raw numbers that are guaranteed to increase. In either case, the
timestamps have microsecond precision if supported by the underlying
platform.

karkhaz commented:
Thanks @tautschnig, I've applied all your suggestions. The code is also now (finally) building on Windows. I've also fixed up everything that clang-format was complaining about, apart from the usual noise from the parse_options strings.

martin-cs commented on src/util/timestamper.h:
```diff
@@ -0,0 +1,92 @@
+  "                              digit zero-padded microsecond counter.\n"     \
+  "                              Timestamp may decrease on daylight saving\n"  \
+  "                              time changes\n"
+#endif
```
Help messages are good.  Verbose help messages are good.  Help messages that are significantly more verbose than all of their friends ... might be better as a slightly more succinct help message and an entry in the manual.

martin-cs reviewed (approved):
Comments but no show stoppers.

martin-cs commented:
@karkhaz : as regards API changes.  For me having one, good, consistent API is the key thing.  Your approach to fixing it sounds reasonable.  Certainly using std functions avoids the usual cross-platform mess that is time keeping.

karkhaz commented:
@martin-cs not sure what you're +1ing on `cbmc_parse_options:1032`, is it @tautschnig's suggestion that I add this option to goto_{analyzer,diff} also? If so that's done.

I've just rebased and made the help text more succinct. Thank you!

martin-cs commented:
@karkhaz It was and yes, done.  Thanks.  This is OK by me once CI passes.

tautschnig commented:
@karkhaz It seems time_stopping.cpp and timer.cpp have been removed from the build, but not actually from the repository?

karkhaz committed:
Replace util/timer* with std::chrono

The old timer API was written in 2004 and is superseded by std::chrono.
All clients of the old API are updated to use std::chrono, and the old
API classes are removed. This is in preparation for a commit that uses
features of chrono that are not present in the old time API.

karkhaz commented:
@tautschnig Ah, thanks! I accidentally destroyed the commit that replaced time{r,_stopping} with std::chrono. I've just re-written it, but I'm afraid I'm working blindly: the build seems to be broken on my machine, I'll see what CI says.

karkhaz commented:
(nice thing about inverted sleep-schedule is that I have the CI server to myself)

karkhaz commented:
@tautschnig I've updated the PR to remove the old time API like before. For some reason, it's failing some of its tests, I'm not sure why---may I have some guidance on how to debug this?
